### PR TITLE
[ML][Data Frame] adding dynamic cluster setting for failure retries

### PR DESCRIPTION
**Pull request by benwtrent**
This adds a new dynamic cluster setting to configure the number of times a non-fatal error is retried by a transform.

The default of 10 is just to keep consistent with what we already have done (nothing special about the number). Minimum of 0, where 0 means it NEVER retries. Maximum of 100 (arbitrary max, but I think 100 is plenty of times).

### CODE REVIEW

**Comment by benwtrent** [comment]
run elasticsearch-ci/2

File: docs/reference/settings/data-frames-settings.asciidoc
Comment (droberts195) [inline docs/reference/settings/data-frames-settings.asciidoc]: I think `data_frames` should be changed to the singular `data_frame`.

There is already a data frame setting, and it begins with `xpack.data_frame` (singular): https://github.com/elastic/elasticsearch/blob/623a5ddac092c3443290580f6b145559e05d64f0/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackSettings.java#L44

(Also, is `xpack.ml.enabled` documented on the ML settings page? If so then maybe `xpack.data_frame.enabled` should be added to this page.)

File: docs/reference/settings/data-frames-settings.asciidoc
Comment (przemekwitek) [inline docs/reference/settings/data-frames-settings.asciidoc]: Just for my education: does the update require restarting the nodes in the rolling fashion or is it truly 'dynamic' in the sense that the nodes will just pick up the new value and start using it right after the update?

File: x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java
Comment (przemekwitek) [inline x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java]: Please finish up the comment.

**Review by przemekwitek — COMMENTED** [review]
Do you think such a change needs tests of any kind?

File: docs/reference/settings/data-frames-settings.asciidoc
Comment (benwtrent) [reply docs/reference/settings/data-frames-settings.asciidoc]: It is truly dynamic, transforms just won't recognize it until they are individually restarted (as the variable is set on start)

**Comment by benwtrent** [comment]
@przemekwitek there are places in the tests where this specific setting would be useful, but those tests are currently muted. I think it can be tested through the integration tests that validate how tasks interact with their failed status. I am planning on updating those tests (and unmuting them) after this PR is merged.

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
run elasticsearch-ci/packaging-sample

**Review by przemekwitek — APPROVED** [review]
LGTM

File: docs/reference/settings/data-frames-settings.asciidoc
Comment (lcawl) [inline docs/reference/settings/data-frames-settings.asciidoc]: Unless other types of data frame settings will be added here, I suggest changing it to:
```suggestion
=== {dataframe-transforms-cap}  settings in Elasticsearch
```

**Review by droberts195 — APPROVED** [review]
LGTM if you could just change the position of the new settings page within the list of X-Pack features.

File: docs/reference/settings/data-frames-settings.asciidoc
Comment (lcawl) [inline docs/reference/settings/data-frames-settings.asciidoc]:
```suggestion
communication requests related to {dataframe-transform} APIs.
```

File: docs/reference/settings/data-frames-settings.asciidoc
Comment (lcawl) [inline docs/reference/settings/data-frames-settings.asciidoc]: Asciidoctor doesn't like attributes in this type of titleabbrev unless you add the following:
```suggestion
[subs="attributes"]
++++
```

File: docs/reference/settings/data-frames-settings.asciidoc
Comment (lcawl) [inline docs/reference/settings/data-frames-settings.asciidoc]: Since neither of these settings seem to be dynamic, I think we should comment out "The dynamic settings..." and the "TIP: Dynamic settings..." content.  
If one of these settings *is* dynamic, we should add "(Dynamic)" after its name (like xpack.ml.max_open_jobs, for example, in https://www.elastic.co/guide/en/elasticsearch/reference/master/ml-settings.html).

File: docs/reference/settings/data-frames-settings.asciidoc
Comment (lcawl) [inline docs/reference/settings/data-frames-settings.asciidoc]: Oops, one more:
```suggestion
<titleabbrev>{dataframe-transforms-cap} settings</titleabbrev>
```

**Review by lcawl — APPROVED** [review]
Documentation LGTM and builds successfully

**Comment by benwtrent** [comment]
run elasticsearch-ci/2

**Pull request merged**
